### ISSUE TITLE
fftw_normxcorr deadlocks in multiprocessing pool

### ISSUE THREAD

**Issue by d-chambers** [eqcorrscan/EQcorrscan#118]
fftw_normxcorr deadlocks when used with a multiprocessing pool, consider this example:

```python
from multiprocessing import Pool, cpu_count

import numpy as np
import obspy

from eqcorrscan.utils.correlate import fftw_normxcorr

cores = cpu_count() if cpu_count() < 4 else 4

# load streams
trace_list = [obspy.read()[0] for _ in range(10)]

# run fftw_normxcorr in serial, this works fine
for trace in trace_list:
    fftw_normxcorr(trace_list[0].data.reshape(1, -1), trace.data, [0])

# run in parallel using multiprocessing pool, this deadlocks
pool = Pool(processes=cores)
args = ((trace_list[0].data.reshape(1, -1), trace, [0]) for trace in trace_list)
results = [pool.apply_async(fftw_normxcorr, args=arg) for arg in args]
pool.close()
# Extract the results when they are done
dist_list = [p.get() for p in results]
# Close and join all the processes back to the master process
pool.join()
```
I am using python 3.6.1 on Ubuntu 16.04

**Comment by calum-chamberlain** [comment]
I think this is because `fftw_normxcorr` will call the threaded C function by default, which works fine in serial but when called multiple time gets confused. At the moment, the `multichannel_normxcorr`, when called with a number of cores, will call `fftw_normxcorr` with threading=False.

Its not perfect, but this is known - there is a note in the arguments for the docs for `fftw_normxcorr`: [here](http://eqcorrscan.readthedocs.io/en/develop/submodules/autogen/eqcorrscan.utils.correlate.fftw_normxcorr.html#eqcorrscan.utils.correlate.fftw_normxcorr). And I will be working on a proper openMP parallel solution over the coming months.

For now I recommend either implementing a serial loop over `fftw_normxcorr` (faster with threading turned on), or a parallel loop without threading in `fftw_normxcorr`. The parallel loop without threading appears to be the fastest implementation so far.

**Comment by d-chambers** [comment]
Ok, I did notice the note on the parameter but I thought it was worth documenting the exact behavior, and how to reproduce it, in an issue. Was there already an issue I missed?

This present a problem in using the normxcorr signature as an interface. I was hoping to make all the functions that use this signature interchangeable. I will need to think more about how to do it in a clean way with this limitation.

**Comment by calum-chamberlain** [comment]
Hmm, I might recommend turning off the threading then. Because they are internal functions I would be happy to have the default for threading changed if that helps? Or maybe the fftw_normxcorr could detect of it was called in parallel... Not sure how to do that though?

**Comment by calum-chamberlain** [comment]
And no, there wasn't an issue you missed, so this is a very good thing to point out and is very appreciated - do you think I should link to this in the docs, or should/could describe it better (if we keep with the same function calls)?

**Comment by d-chambers** [comment]
It might be good to link in the docs. Just in case people are getting unexplained deadlocks. If you can turn off threading by default it would make it much less likely someone stumbles into this.

**Comment by calum-chamberlain** [comment]
Hey @d-chambers has #139 taken care of this, or is the method that generates the deadlock still accessible? e.g. Is it still possible to call the threaded C-func and Python multiprocessing at the same time?
If it isn't then I will close and link to that PR.

**Comment by d-chambers** [comment]
It shouldn't be possible any longer to get to the deadlock.

**Comment by calum-chamberlain** [comment]
Closed by #139
